woocommerce/woocommerce
# Do not dismiss the error snackbar automatically

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds the code to don't dismiss the error snackbar automatically.

When the user creates a product and clicks `Publish`, `Preview` and `Save draft`, a snackbar will be shown if the form has errors.

![Screenshot 2024-06-06 at 3 24 41 PM](https://github.com/woocommerce/woocommerce/assets/1314156/552deca3-1f3c-4cc4-b187-d2a25fb85412)

Closes #46765.

A few considerations about the ACs:
- Now it's not possible to add a non-dismissible snackbar (it's [a known thing](https://github.com/WordPress/gutenberg/issues/45824), probably on purpose). This is why we use the prop `explicitDismiss` that also adds an `X` to close the snackbar.
- One of the original ACs mentioned that if the user was viewing the tab with the error, the snackbar would be shown for one second. However we show the snackbar longer by default. I modified that AC to show the snackbar for the period shown by default. 

```[tasklist]
### ACs
- [ ] Don't dismiss the error snackbar automatically.
- [ ] Dismiss the snackbar if the user is seeing the tab. The error will disappear automatically after a few seconds.
- [ ] If there are more error messages, we'd stack them.
```



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure `Try the new product editor (Beta)` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`
2. Go to `Products > Add new` and add a Summary text.
3. Press `Save draft`, `Preview` and `Publish`. After pressing these buttons you will see 3 snackbars with the message `Product name is required.`.
4. Since you are in the `General` tab and the pricing is visible, the snackbar will disappear automatically.
5. Go to the `Shipping` tab.
6. Repeat step 3. The same snackbar will be shown and only will be dismissed after pressing the `X` next to the test.
7. Go to the `Variations` tab and create some.
8. Edit a single variation.
9. Add a `Note` and try to `Update` the variation.
10. You will see a snackbar with the message `Regular price is required.`, it will dismiss automatically. 
11. Go to the `Inventory` tab and press `Update`.
12. The same snackbar will be shown with an `X` at the end to dismiss it.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
